make the CGIR goto-based and fix exception-handling bugs

zerbina opened the pull request:
## Summary

* extend the MIR's design with goto-based control-flow primitives
* implement the new control-flow primitives for the CGIR
* make the C code generator use the new control-flow primitives
* (C backend) fix multiple bugs with `getCurrentException` returning
  the wrong exception
* (C backend) fix exceptions being leaked when aborting a `finally`
  section

### Details

The main goals of replacing the higher-level control-flow primitives in
the MIR with lower-level, goto-based ones are to:
* be able to express more complex control-flow
* make control-flow-related transformations easier and/or possible
* reduce the cost of computing the data-flow graph
* make evolving the language and MIR easier

A guiding principle was that the design should allow for enough freedom
to generate efficient code with all three code generators (which differ
significantly in control-flow-related capabilities).

To be able to implement this incrementally, as a first step, the new
control-flow primitives are only added to the CGIR, with only `cgen`
actually supporting them. So that `jsgen` and `vmgen` continue to work
without change, the legacy primitives are kept, with `cgirgen_legacy`
(a full copy of the original `cgirgen`) continuing to produce the
legacy CGIR.

### MIR

An in-depth description of syntax and behaviour is added to
`docs/mir.rst`. The gist of the design is that:
* terminators of basic-blocks become explicit (an exception being `If`)
* jump targets are directly specified on jump/fork-like operations
  (e.g., `Goto`, `CheckedCall`, etc.)
* intercepted control-flow (via `finally` sections) is also specified
  on jump/fork-like operations
* `finally` sections are *not* required to be duplicated for each
  unique control-flow passing through them
* jump targets and join points are identified by *labels*, which are
  integers IDs unique within a body

As a consequence of this design, there's no more statement-in-statement
nesting.

### CGIR & Translation

The CGIR implements all the new control-flow primitives from the MIR,
with the only difference being that the CGIR's `ContinueStmt` doesn't
track the possible exits of a `Finally` section (it's information
irrelevant to the code generators).

`cgirgen` handles translation of the old-style primitives still used by
the MIR to the new-style ones. While not strictly needed at the moment,
translation is disabled of intra-procedure unreachable code is
disabled, meaning that unreachable code is effectively dropped. This is
so that `Finally` sections that never exit normally (e.g.,
`finally: return`) can be ensured to work.

### C code generation

The CGIR is first translated into a small, low-level, C-specific IR
(all implemented in `ccgflow`). Here, the focus is on figuring out what
C code to produce for `Finally` sections; unnecessary gotos are also
eliminated.

#### Finally Handling

The previous strategy for `finally` was to:
* emit an error-state considering version of the `finally` at the end
  of the `try`
* duplicate the body of all enclosing `finally`s at `break`s and
  `return`s

Now, only a single version of the `finally` section is ever emitted. If
the `finally` section has more than one possible exit, a run-time
dispatcher is used, like so:
```c
{
  NI32 Target0_;
  L1_1_:
  Target0_ = 0;
  goto L1_;
  L1_2_:
  Target0_ = 1;
  L1_: // never jumped to from the outside
  NI32 oldNimErr0_ = *nimErr_; *nimErr_ = NIM_FALSE; // temporarily disable error mode
  // ... body of the finally ...
  *nimErr_ = oldNimErr0_;
  switch (Target0_) {
  case 0: goto L2_;
  case 1: goto L3_;
  }
}
```

Depending on the context and complexity of the code in-between, an
optimizing C compiler is able to eliminate the dispatcher and
`Target0_` assignment by directly inlining the body at a `goto L1_1_;`.

For the common case of a `Finally` section having two exits - one for
exceptional control-flow cases, and one for the -, a
`if (NIM_UNLIKELY(*nimErr_)) goto <error_label>; goto <non_error>` is
used instead of a full dispatcher. This mirrors how the exit of a
`finally` section previously looked like.

The idea behind not duplicating `Finally` sections is that it reduces
the amount of C code the compiler has to output. Nonetheless, the IR
from `ccgflow` is flexible enough to use the duplication strategy,
should this be needed again.

#### Exception Handling

Whether control-flow crosses the boundary of a (i.e., leaves it) is
encoded in the MIR/CGIR via the `Leave` item in target lists. `ccgflow`
uses this information to make sure that leaving a `Finally` via
unstructured control-flow properly aborts the active exception (if any)
and that leaving an `Except` properly pops the handler.   

#### Other

The IR from `ccgflow` then drives the rest of C code generation. Since
there's no more statement-in-statement nesting, recursion reduces
significantly.

A side-effect of the flatter representation is that less C scopes are
used (only `if`, loops, `finally`, and `except` use C scopes), which
seems to harm C compiler optimization.

Support for the legacy control-flow primitives is fully removed.

### Exception Handling Runtime

The runtime part of exception handling for the C target is partially
overhauled. For compatibility with the `csources` compiler, the new
version is guarded behind the `nimskullNewExceptionRt` condsym.

A distinction is now made between an *in-flight* exception and the
*current* exception. Previously, both were treated as being one and the
same.

An exception becomes *in-flight* once it is raised (`raiseException2`),
and stops being in-flight when it is caught (`nimCatchException`).
While in-flight, an exception can be aborted (`nimAbortException`), by
breaking out of an intercepting `finally`. On being caught by a
handler, the exception is associated with the handler and the handler
pushed to the handler stack. When control-flow leaves the handler
(`nimLeaveExcept`), the handler is popped from the stack.

The new runtime fixes the `currentException` not being updated when
breaking out of a `finally` or raising from within an `except` handler.
In addition, an exception caught by a handler can now be raised again
within the handler (e.g., `except CatchableError as e: raise e`)
without reference cycles being introduced.

zerbina committed:
introduce `cgirgen_legacy` and associated procedures

It's a full copy of `cgirgen`, intended to keep `jsgen` and
`vmgen` working without larger changes. Once both code generator are
updated, `cgirgen_legacy` can be removed again.

zerbina committed:
mir: document the new control-flow primitives

Documentation for the old ones is kept for now. It'll be removed once
the they're gone too.

zerbina committed:
jsgen, vmgen: ignore the new CGIR nodes

zerbina committed:
cgir: add the new control-flow statement/nodes

They use the syntax described by `mir.rst`. The legacy node kinds are
kept in order for `jsgen` and `vmgen` to continue working.

zerbina committed:
cgirgen: produce new-style CGIR

`cgirgen` now emits goto-style CGIR adhering to the structure laid out
in the MIR documentation.

A minor difference compared to the MIR specification is that `cgirgen`
doesn't add potential follow-up targets to `cnkContinueStmt`s. It would
only introduce unnecessary complexity at the moment, the code
generators don't need this information.

In addition, a simple form of eliminating unreachable code is added.
This allows for making sure that `cgen` properly handles the case where
a `finally` is the final jump target of a jump.

zerbina committed:
add the `ccgflow` module

The module implements the translation of CGIR to a mini, control-flow
focused abstract machine. This is intended to decouple control-flow
related decision making and optimization from the rest of C code
generation.

zerbina committed:
ccgstmts: integrate `ccgflow`

zerbina committed:
cgir: update labels when merging bodies

Same as with locals, the IDs of labels in the source body need to be
updated, otherwise they would clash with the existing ones.

zerbina committed:
cgen: account for checked call nodes

The additional trailing node needs to be ignored by logic only
interested in the arguments.

zerbina committed:
cgen: update call exit handling

The node from the target slot needs to be passed along to `raiseExit`.

zerbina committed:
cgen: update `raiseInstr`

Where to jump to is now specified by node in the target slot of the
call. The node is passed along to `raiseInstr`, where the jump target
is extracted. No jump target slot (when the call is not a
`cnkCheckedCall`) is interpreted as unreachable.

zerbina committed:
cgen: update code generation for nodes

* `cnkBlockStmt`, `cnkTryStmt`, `cnkBreakStmt`, `cnkReturnStmt`, and
  `cnkStmtList` are all part of the legacy node; support for them is
  removed
* `cnkRepeatStmt` doesn't need a dedicated procedure anymore
* `cnkEnd` only closes a C scope
* remove the `withinLoop` counter

zerbina committed:
cgen: replace `withinTryWithExcept` detection

There are no `try` blocks anymore, so it doesn't work. Whether
assigning the result of a checked call is a problem (because it'd be
incorrectly observable) is now detected by analysing the target (or
target list) of the call.

zerbina committed:
cgen: update exception code gen

* emission of the `goto` for raise statements is handled by `ccgflow`
* `cnkTryStmt` is replaced by `cnkExcept` (a single exception handler)
* try/except/finally tracking state is removed

zerbina committed:
cgen: make `reportObservableStore` work again

The `nestedTryStmts` counter is gone. Instead, the target (list) is
analysed to detect whether the an exception handler or finally is
visited.

zerbina committed:
ccgstmts: update `case` statement code generation

Since `cnkCaseStmt` is a goto-based dispatcher now, code generation
becomes simpler, especially for string case statement, as those already
used gotos.

zerbina committed:
cgen: remove obsolete `TBlock` data and routines

Blocks no longer have lables attached or need to keep track of
try-statement-related information.

zerbina committed:
ccgstmts: implement new exception handling

The three new, not-yet-implemented additions to the runtime are:
* `nimCatchException`
* `nimAbortException`
* `nimLeaveExcept`

The latter to are used to implement the `opAbort` and `opPopHandler`
instructions, `nimCatchException` is emitted at the start of an
`except` section.

zerbina committed:
cgir: use `cnkEnd` to mark end of except section

This is going to make code generation for except sections much easier,
especially for the JS code generator. For the C code generator, an
`except` section is now wrapped in a scope again.

zerbina committed:
ccgflow: implement exception-cleanup handling

The `Leave` operations are now also part of the jump path, where
they're effectively treated like finallys. For `finally`s, an `opAbort`
instruction is emitted for the cleanup, for `except`s a `opPopHandler`.

zerbina committed:
ccgstmts: make sure the `BeforeRet_` label exists

zerbina committed:
ccgflow: eliminate unnecessary jump

Redirected labels weren't considered for single-target gotos, leading
to unnecessary indirections. This only improves readability of the C
code, as it's trivial for a C compiler to optimize the goto away.

zerbina committed:
excpt: implement the new exception runtime

Since there's overlap with the legacy runtime procedures, the
conditional symbol `nimskullNewExceptionRt` is required to identify
whether the new runtime is used.

zerbina committed:
excpt: annotate problematic handling of `currentException`

The manual states that `getCurrentException` only returns the exception
caught by an `except` handler, *not* in-flight exceptions.

zerbina committed:
ccgstmts: better code generation with `--panics:on`

Omitting the jump after calls of the `.noreturn` compiler runtime
procedures produces much worse machine code. If available, the control-
flow path is marked as unreachable using a C compiler built-in,
otherwise a jump-to-exit is used.

zerbina committed:
ccgflow: document why redirected labels cannot be omitted

zerbina committed:
tests: enable `tfinally6.nim`

It *finally* works ;)

zerbina committed:
ccgstmts: wrap `finally` sections in a scope

For complex `finally` sections, this prevents locals bleeding into the
above block.

zerbina committed:
used dedicated node pair for loops

A loop is now started by a `cnkLoopJoinStmt` and ends in a
`cnkLoopStmt`. Overloading `cnkEnd` to mean terminator+join and
terminator makes local reasoning about `cnkEnd` impossible,
unnecessarily complicating analysis.

zerbina committed:
tests: extend the test suite for `except`

All added tests previously failed with the C backend.

zerbina committed:
fix "declared but not used" hint

For some reason, the `isFilled` routine in `cgirgen_legacy` is
considered not used, even though it should. The compiler doesn't
complain about it missing, so removing it seems fine.

zerbina committed:
cgirgen: fix too aggressive unreachable code elimination

Disabling translation within unscoped contexts can be a problem with
how destructors are currently injected, therefore unscoped contexts are
always translated for now.

zerbina committed:
ccgflow: properly take `isFull` into account

If the statement list doesn't constitute the whole body, trailing joins
cannot be turned into the exit join point.

zerbina commented:
There's a pre-existing bug with exception handling that still persists:
```nim
try:
  try:
    raise CatchableError.newException("a")
  finally:
    var e = CatchableError.newException("b")
    try:
      try:
        raise e
      finally:
        try:
          raise e 
          # `activeException.up` now points to `activeException`; the stack
          # of in-flight exceptions is corrupted
        except:
          discard
    except:
      discard
    # both the current and active exception are now nil, but error mode is still
    # active (correctly so). The next exception handler is going to crash

except CatchableError as e:
  doAssert e.msg == "a" 
```
This snippet crashes (previously the wrong exception reached the final handler).

Personally, I think the best solution here is to panic when an exception that's already in-flight is raised again, although that's something better implemented in a separate PR.

saem reviewed (approved):
Did a quick skim and then went through the commits. Looks good, really excited to see the remaining backends come over.

disruptek commented:
This branch exhibits some new compiler crashes in the cps tests when run without `panics:on`, though I haven't looked too closely and it's possible that cps shares some blame here.  Nonetheless, the improvements are not fully bug-compatible with tip. :shrug:

zerbina committed:
cgirgen: further restrict unreachable code elimination

Translation is enabled again at all non-`except`, non-`finally` join
points, effectively disabling unreachable code elimination.

zerbina commented:
@disruptek Thanks for testing the branch with CPS. I was able to isolate and fix the problem.

---

The problem was elimination of unreachable being too aggressive giving the current. Take the following MIR:
```
scope:
  try:
    block L0:
      if cond:
        scope:
          return
      scope:
        return
    def x = construct (...)
    if cond:
      scope:
        return
  finally:
    =destroy(name x)
```
which is currently produced for:
```nim
if cond:
  return
else:
  return

var x = SomeType(...)
if cond:
  return
```

The first two returns are part of a *scoped* context (they're part of a `scope` block and there's no `if`, `repeat`, `try`, etc. in-between), and thus to-CGIR translation of MIR statements following them is disabled. Translation is thus inactive at the end of the `L0` block, meaning that everything up to the `finally` was discarded.

In theory, this is correct, but since the `def x` wasn't translated, information about the local is missing when `=destroy(name x)` (which is reachable) is translated. For the time being, translation is forcefully enabled again at every non-`except`, non-`finally` join point. Once the MIR itself uses the goto-based control-flow representation, this workaround becomes obsolete.

alaviss commented:
> Personally, I think the best solution here is to panic when an exception that's already in-flight is raised again, although that's something better implemented in a separate PR.

We could use an alive count for this, if the current exception is the same as in-flight head, then just increase the alive count so we could pop it once we leave the handler that raised it.

zerbina commented:
> We could use an alive count for this, if the current exception is the same as in-flight head, then just increase the alive count so we could pop it once we leave the handler that raised it.

Hm, I'm not sure I understand of how that would help. Since the in-flight stack is singly-linked list, pushing an exception to the stack that's already part of the stack corrupts the stack.

alaviss commented:
What I meant is something like this

```nim
proc pushException(e: Exception) =
  if inflight == e:
    e.extraRef.inc
  else:
    e.up = inflight
    inflight = e

proc popException() =
  if e.extraRef > 0:
    dec e.extraRef
  else:
    inflight = inflight.up
```

zerbina commented:
Ah, I see what you mean, thanks for the example.

Unfortunately, this would only work for the narrow case where the in-flight exception at the head of the stack is raised again. Consider:
```nim
var e1, e2: CatchableError = ...
try:
  raise e1
finally:
  try:
    raise e2
  finally:
    try:
      raise e1
    finally:
      try:
        raise e2
```
This would still fail.

Even incrementing the `extraRef` counter when the exception is not the in-flight head wouldn't help: the meaning of the stack would be ambiguous. `e2 -> e1 -> e2 -> e1` and `e2 -> e1 -> e1 -> e2`, and `e2 -> e2 -> e1 -> e1` couldn't be distinguished from each other.

alaviss commented:
True that, maybe we should just give up on linked list and use a seq :p

Or maybe a linked list of arrays so we can have a guaranteed stack count then allocate extra as needed

```nim
type
  ExceptionStack = ptr object
    data: array[<a sensible number>, ref Exception]
    len: int
    prev: ExceptionStack
```

zerbina commented:
/merge

I think the issue with corrupting the in-flight exception stack should be fixed separately. It's pre-existing, so no regression is introduced.
